SWIPs/swip-data-types-descriptions.md

```diff
@@ -0,0 +1,90 @@
+---
+SWIP: <to be assigned>
+title: Type definitions for protocol messages
+author: Louis Holbrook @nolash <dev@holbrook.no>
+status: Draft
+type: General
+created: 2019-08-02
+---
+
+## abstract
+
+This SWIP defines pseudo data types for use in Swarm protocol specifications.
+
+## motivation
+
+To describe protocols in our specifications, we are in need of a strongly defined way of describing data types and their serializations to RLP.
+
+RLP does not have "types" as such, just self-describing data units whose length is serially interpreted. Hence it is for example not possible to define a number value range merely in terms of RLP constructs.
+
+## Specification
+
+This document uses the `ABNF` convention for defining data types, as
+described in the IETF RFC 5234. [@IETF:ABNF]
+
+### primitives
+
+The basic identifiers are defined as follows:
+
+``` {numbers="none"}
+UINT64MAX   = 18446744073709551615
+UINT32MAX   = 4294967295
+UINT16MAX   = 65535
+UINT8MAX    = 255
+UINT64      = %d0-UINT64MAX
+UINT32      = %d0-UINT32MAX
+UINT16      = %d0-UINT16MAX
+UINT8       = %d0-UINT8MAX
+
+BOOL        = BIT
+
+TIMESTAMP   = UINT32
+```
+
+### rule extension for list definitions {#abnf-list-definition}
+
+The construct `LIST` explicitly defines a list of elements. Elements may
+be of any type, and are separated by spaces. This identifier has
+implications for the RLP encoding of the data, as described in [3].
+
+### general types
```

Comment:
Is this section from latex? it could probably be formatted in a more meaningful way in markdown.

Reply:
Yes, as far as I know, markdown doesn't have any notion of definition lists. This kind of thing is one of the reasons why I don't like fast-food solutions like markdown in the first place.

How should we represent definition lists - any suggestions?

Reply:
I guess as a table?

Reply:
Tables work just fine in markdown, here are some examples from eth2 networking spec of how this can be done https://github.com/ethereum/eth2.0-specs/blob/02bb92e71455adaa7da101563a6c367efe9e1cc7/specs/networking/p2p-interface.md

Reply:
actually it seems like their definition lists are like this:

* `key` value